Technote about credentials (#713)
ropensci/roweb2: content/technotes/2020-07-07-github-pat-credentials.md

---
slug: github-pat
title: 'A better way to manage your Github personal access tokens'
date: '2020-07-07'
author:
  - Jeroen Ooms
tags:
  - R
  - package
  - gert
  - credentials
  - git
  - github
---

> We have been working hard behind the scenes on the upcoming release of our new git package named [gert](https://docs.ropensci.org/gert), a joint effort from rOpenSci and the Tidyverse team. One of the main features of gert is the out-of-the-box authentication mechanism, which is provided via the new [credentials](https://docs.ropensci.org/credentials/articles/intro.html) package.

Reviewer: Nitpick: I write about Git a lot too in usethis and have decided to try to say capital-G "Git" because that seems to be the official capitalization. I also acknowledge this is utterly meaningless. I won't bring it up again.

> Among other things, the credentials package makes it possible to save and load https authentication details from the [git credential store](https://git-scm.com/docs/git-credential-store), which is part of the official command-line git. Thereby credentials are automatically shared between command line git and the gert package, while safely stored by your operating system preferred password manager.

Reviewer: ```suggestion
Among other things, the credentials package makes it possible to save and load https authentication details from the [git credential store](https://git-scm.com/docs/git-credential-store), which is part of the official command-line git. Thereby credentials are automatically shared between command line git and the gert package, while safely stored by your operating system's preferred password manager.
```

> In this post we show how you can take this one step further, and use the credentials package to save your `GITHUB_PAT` in the git credential store. This way you can authenticate with the GitHub API using the same token that is used for HTTPS remotes in git and gert. This is convenient for users, and also provides package authors with a mechanism to prompt the user for credentials, without having to take responsibility for managing tokens.

Reviewer 2: My comment is maybe more about the package docs, or for a footnote: could you explain how credentials relate to https://github.com/r-lib/keyring? "Maybe you were looking for blabla" (and maybe other packages add to the confusion?)

Author: Hmm I think that's beyond the scope of the article. Keyring interfaces to the local password manager, but not via git.

## What is a Personal Access Token

> GitHub allows you to generate [Personal Access Tokens](https://github.com/settings/tokens), which you can use instead of your password when authenticating over HTTPS, both for git remotes and the GitHub API. Major advantages are:

Reviewer 2: Especially since you say responsible users have to use 2FA :stuck_out_tongue_winking_eye:

>  - If you have enabled 2FA, you need a PAT to authenticate programmatically.

Reviewer: ```suggestion
 - If you have enabled 2FA, you must use a PAT to authenticate programmatically. You have no choice.
```

>  - A PAT can easily be revoked or updated. You can even automate this via the Github API.

Reviewer: Updated? I don't think that's true.My impression is that if you need to get the actual PAT or update its scopes, you have to start over. But the ability to revoke individual PATs seems like a really important point.

Author: Hmm I thought the Github API would let you generate and revoke PATs? 🤔

>  - You can generate many PATs with specific permissions, giving you fine grained security control.

Reviewer: ```suggestion
 - You can generate many PATs with specific permissions, giving you fine-grained security control.
```

In conclusion, if you are a responsible GitHub user, you have enabled 2FA on your account, and you only ever enter your main password when authenticating on the GitHub website. Everywhere else you should be using a PAT, preferably one that only has the permissions it needs.

## Storing and loading your GitHub PAT

> Most R packages that interact with the GitHub API expect that your PAT is stored in the environment variable `GITHUB_PAT` in the R process. But how does it get there? The credentials package provides a function that will set this environment variable:

Reviewer: You might need to spell out what most of us do: define the env vars in `~/.Renviron`.

```r
# Tries to set the GITHUB_PAT environment variable
credentials::set_github_pat()
## TRUE
```

This function calls out to the [git credential store](https://git-scm.com/docs/git-credential-store) to get a suitable token for the `github.com` domain. If no token is available yet, the git credential manager will then prompt the user to enter one. What this looks like depends on your operating system and [credential helper](https://git-scm.com/docs/gitcredentials) configuration.

![token](https://i.imgur.com/cnJcRmw.png)

> 

Author: I'll try to improve that. But note that there is a bit of magic going on here because this prompt is invoked not by R, but by `git` when it calls back to the credential store. In some front-ends we don't have full control over the appearance of the password prompt (but in this example we do).

> The `set_github_pat` function returns `TRUE` when it succeeded in setting the `GITHUB_PAT` environment variable, and `FALSE` if not. Packages that call `set_github_pat()` to let the user authenticate, can check the return value to determine if authentication was successful.

Reviewer: ```suggestion
The `set_github_pat()` function returns `TRUE` when it succeeds in setting the `GITHUB_PAT` environment variable, and `FALSE` if not. Packages that call `set_github_pat()` to let the user authenticate, can check the return value to determine if authentication was successful.
```

> Once a working PAT has been stored in the git credential store, it can automatically be loaded in another R session by calling `set_github_pat()` again. The token is automatically validated, and if it still works, the `GITHUB_PAT` environment variable is set without the user having to do anything.

Reviewer: I didn't realize credentials validates the token! I will have to look at how you do this. Usethis also does this, but I'd love to delegate to credentials.

> And here is the best part: because the token is stored under the `github.com` domain in the credential store, both gert and command line git will automatically attempt to authenticate with this token when fetching/pushing Github HTTPS remotes.

Reviewer: Indeed!

## Managing the credential store

How long does the credential store remember your token? This depends on which [credential helper](https://git-scm.com/docs/gitcredentials) is in use. On all systems, git includes at least the following helpers:

- `cache`: *Cache credentials in memory for a short period of time. See [git-credential-cache](https://git-scm.com/docs/git-credential-cache) for details.*

> - `store`: *Store credentials indefinitely on disk. See [git-credential-store](https://git-scm.com/docs/git-credential-store) for detail*

Reviewer: The first link, way at the top, goes here. And a basic reading of the docs for credential store doesn't make it seem any more secure than `~/.Renviron` to me. It's only when we start talking about the macOS keychain and Windows credential manager that it starts to seem meaningfully different.

Author: You are right, the first link should go to https://git-scm.com/docs/gitcredentials.

The main message here is that the responsibility of safely storing the token is moved to the user, or at least the local git credential configuration.

However on MacOS and Windows, git actually defaults to custom credential helpers that use the OS password manager. On MacOS this is called keychain and on Windows this is [git credential manager for Windows](https://github.com/microsoft/Git-Credential-Manager-for-Windows). These credentials helpers can store your credentials indefinitely.

```r
credentials::credential_helper_get()
## [1] "osxkeychain"
```

The credentials package includes a few more utility functions to help you interact with the credential store. To manually load credentials for a given domain use `git_credential_ask`:

```r
credentials::git_credential_ask("https://github.com")
## $protocol
## [1] "https"
## 
## $host
## [1] "github.com"
## 
## $username
## [1] "token"
## 
## $password

> ## [1] "aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d"

Author: Only one way to find out

```

Use `git_credential_forget` to explicitly remove a credential:

```r
# Drop GITHUB_PAT credential from store
credentials::git_credential_forget('https://token@github.com')
```

Alternatively, if you want to switch to another PAT, use `set_github_pat(force_new = TRUE)`. This will automatically drop any existing PAT from the credential store, and always prompt the user to enter a new one.

```r
# Drop existing GITHUB_PAT and ask for a new one
credentials::set_github_pat(force_new = TRUE)

> ```

Reviewer 2: there's no conclusion but I suppose you know that :grin:
